CRM/ManualDirectDebit/Common/MandateStorageManager.php

```diff
@@ -30,19 +30,35 @@ public function assignContributionMandate($contributionId, $mandateID) {
   /**
    * Assigns dependency between recurring contribution and mandate
    *
-   * @param $contributionId
+   * @param $contributionRecurId
    * @param $mandateId
    */
-  public function assignRecurringContributionMandate($contributionId, $mandateId) {
-    $rows = [
-      'recurr_id' => $contributionId,
-      'mandate_id' => $mandateId,
-    ];
+  public function assignRecurringContributionMandate($contributionRecurId, $mandateId) {
+    $existingMandateId= CRM_ManualDirectDebit_BAO_RecurrMandateRef::getMandateIdForRecurringContribution($contributionRecurId);
+    if ($existingMandateId && $existingMandateId == $mandateId) {
+      return;
+    }
+
+    if ($existingMandateId) {
+      $activityType = 'update';
 
-    CRM_ManualDirectDebit_BAO_RecurrMandateRef::create($rows);
+      $mandateReferenceId = CRM_ManualDirectDebit_BAO_RecurrMandateRef::getMandateReferenceId($existingMandateId, $contributionRecurId);
+      CRM_ManualDirectDebit_BAO_RecurrMandateRef::create([
+        'id' => $mandateReferenceId,
+        'recurr_id' => $contributionRecurId,
+        'mandate_id' => $mandateId,
+      ]);
+    }
+    else {
```

Comment:
May be cleaner to only have one call to `CRM_ManualDirectDebit_BAO_RecurrMandateRef::create()`.

```php
$activityType = 'create';
$params = [
  'recurr_id' => $contributionRecurId,
  'mandate_id' => $mandateId,
];

if ($existingMandateId) {
  $activityType = 'update';
  $mandateReferenceId = CRM_ManualDirectDebit_BAO_RecurrMandateRef::getMandateReferenceId($existingMandateId, $contributionRecurId);
  $params['id'] = $mandateReferenceId;
}

CRM_ManualDirectDebit_BAO_RecurrMandateRef::create($params);
$activity = new CRM_ManualDirectDebit_Hook_Post_RecurContribution_Activity($contributionRecurId, $activityType);
$activity->process();
```